bluesky/plans.py

```diff
@@ -33,93 +33,47 @@ class PlanBase(Struct):
     """
     subs = Subs({})
 
-    @property
-    def md(self):
-        """
-        a metadata dictionary, passed to the 'open_run' Message
-        """
-        self._md['plan_type'] = type(self).__name__
-        self._md['plan_args'] = {field: repr(getattr(self, field))
-                                 for field in self._fields}
-        for category, objs in self._objects.items():
-            # e.g., self.md['motors'] = ['motor1', 'motor2']
-            if isinstance(objs, Iterable):
-                self._md[category] = [obj.name for obj in objs]
-            else:
-                obj = objs
-                self._md[category] = obj.name
-        return self._md
+    def get_metadata(self):
+        return {}
 
     def __iter__(self):
         """
         yields messages
         """
-        # Some metadata is compute in _pre, so it must be done before
-        # the RunStart is generated by open_run, below.
-        yield from self._pre()
-
-        # If the plan has a 'subs' attribute, yield a 'subscribe' Msg for
-        # each one. The RunEngine will remove these at the end in the
-        # event of termination.
-        tokens = set()
-        subs = normalize_subs_input(getattr(self, 'subs', {}))
-        for name, funcs in subs.items():
-            for func in funcs:
-                token = yield Msg('subscribe', None, name, func)
-                tokens.add(token)
-
-        yield Msg('open_run', **self.md)
-        for flyer in self.flyers:
-            yield Msg('kickoff', flyer, block_group='_flyers')
-        yield Msg('wait', None, '_flyers')
-        yield from self._gen()
-        for flyer in self.flyers:
-            yield Msg('collect', flyer, block_group='_flyers')
-        yield Msg('wait', None, '_flyers')
-        yield Msg('close_run')
-        for token in tokens:
-            yield Msg('unsubscribe', None, token)
-        yield from self._post()
+        md = self.get_metadata()
+        md['plan_args'] = {field: repr(getattr(self, field))
+                           for field in self._fields}
+        flyers = self.flyers
+
+        main_with_flyers = fly_during(self._main_gen(), self.flyers)
```

Comment:
I am pretty convinced that this only works with strict coroutines, not generators, which means any plan that used this could not be list-ified. This issue is likely to come up a lot as we develop fancier wrappers, so perhaps we just need to commit to coroutines and provide better way to list-ify plans.